### DISCUSSION TITLE
HPIPM solver interface (MATLAB)

### DISCUSSION THREAD

**Discussion by jeroenvanhegelsom** [casadi/casadi#3277] [Q&A] [answered]
Hi,

I'm trying to use the (recently included) HPIPM interface of CasADi in MATLAB to solve a QP that I previously solved using qpOASES. I first tried to understand the required structure from the documentation:

![image](https://github.com/casadi/casadi/assets/70114329/8d784fa6-c9f9-422f-8492-2a5eca97e7c1)

However I didn't understand it immediately. A recommendation from another post in this forum  ([https://github.com/casadi/casadi/discussions/3206](url)) was to look at the python example, but when implementing the exact same problem:

![image](https://github.com/casadi/casadi/assets/70114329/65f0f596-355a-468e-83a6-eac64f910ad3)

which I can't get to work in MATLAB. When supplying the structure of A (passing `nx`, `nu`, `ng` and `N` as options), this yields the error 

`.../casadi/interfaces/hpipm/hpipm_interface.cpp:269: Assertion "nu.size()==N_+1" failed`,

while the length of `nu` in the python example is `N` (which makes more sense to me as well). When not passing the structure in the options struct, the structure (`Ak`, `Bk`, `Ck`, `Dk`) is not identified correctly and the error displayed is

`.../casadi/interfaces/hpipm/hpipm_interface.cpp:317: Assertion "(A_ + total).nnz() == total.nnz()" failed`.

Is there any other example for MATLAB specifically or has anyone been able to use the HPIPM interface in MATLAB already? Or does anyone know what I'm doing wrong?

Thanks in advance,

Jeroen

**Comment by jgillis** [comment, marked as answer]
Could this be a Matlab translation error? Here is the code I obtained which does not raise an error:

```matlab
import casadi.*

A=[1, 0.2, 1, -1, 0, 0, 0, 0, 0, 0, 0, 0;
 -0.1, 0.4, 0, 0, -1, 0, 0, 0, 0, 0, 0, 0;
 0.3, 0.2, 0, 0, 0, -1, 0, 0, 0, 0, 0, 0;
 2, 0, 0.3, 0, 0, 0, 0, 0, 0, 0, 0, 0;
 1, 1, 0.4, 0, 0, 0, 0, 0, 0, 0, 0, 0; 
 0, 0, 0, 1, 4, 2, 1, 0.3, -1, 0, 0, 0; 
 0, 0, 0, 3, 1, 0, 1, 0.2, 0, -1, 0, 0; 
 0, 0, 0, 1, 1, 1, 1, 1, 0, 0, 0, 0;
 0, 0, 0, 0, 0, 0, 0, 0, 2, 4, 0, -1; 
 0, 0, 0, 0, 0, 0, 0, 0, 2, 3, 1, 0;
 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 3];

A = DM(sparse(A));

H=[7, 0, 0.2, 0, 0, 0, 0, 0, 0, 0, 0, 0; 
 0, 7, 0.3, 0, 0, 0, 0, 0, 0, 0, 0, 0; 
 0.2, 0.3, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0; 
 0, 0, 0, 3, 0, 0, 0, 1, 0, 0, 0, 0; 
 0, 0, 0, 0, 2, 0.1, 0, 0.7, 0, 0, 0, 0; 
 0, 0, 0, 0, 0.1, 1, 0, 1, 0, 0, 0, 0; 
 0, 0, 0, 0, 0, 0, 1, 0.1, 0, 0, 0, 0; 
 0, 0, 0, 1, 0.7, 1, 0.1, 2, 0, 0, 0, 0; 
 0, 0, 0, 0, 0, 0, 0, 0, 6, 0, 1, 0; 
 0, 0, 0, 0, 0, 0, 0, 0, 0, 6, 0, 0; 
 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 4, 0; 
 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 9];

H = DM(sparse(H));

nx = [2,3,2,1];
nu = [1, 2,1];
ng = [2, 1, 1, 1];
N = 3;


options = struct;
options.hpipm.iter_max = 100;
options.hpipm.res_g_max = 1e-10;
options.hpipm.res_b_max = 1e-10;
options.hpipm.res_d_max = 1e-10;
options.hpipm.res_m_max = 1e-10;

solver = conic('solver', 'hpipm', struct('a',A.sparsity(), 'h', H.sparsity()),options);

g = [1;1;0.2;0.4;1;0.5;0.3;1;0.6;1;1;0.7];
lbg =[0;0;0;-2;-2;0;0;-2;0;-2;-2];
ubg = [0;0;0;2;2;0;0;2;0;2;2];
lbx = [0.5;0.2;-1;-1;-1;-1;-1;-1;-1;-1;-1;-1];
ubx = [0.5;0.2; 1; 1; 1; 1; 1; 1; 1; 1; 1; 1];

sol = solver('a',A,'h',H,'lba',lbg,'uba',ubg,'g',g,'lbx',lbx,'ubx',ubx)
```

**Reply by jeroenvanhegelsom** [reply]
Thanks a lot! 
I was missing that the matrices should first be converted to their sparse form before passing their sparsity into the solver, and now it works. I guess this was also the reason for the error occurring at the assertion about nonzeros in A.
